lib_standard_app/io.c

```diff
@@ -153,22 +153,27 @@ WEAK int io_recv_command()
     return ret;
 }
 
-WEAK int io_send_response_pointer(const uint8_t *ptr, size_t size, uint16_t sw)
-{
-    return io_send_response_buffer(&(const buffer_t){.ptr = ptr, .size = size, .offset = 0}, sw);
-}
-
-WEAK int io_send_response_buffer(const buffer_t *rdata, uint16_t sw)
+WEAK int io_send_response_buffers(const buffer_t *rdatalist, size_t count, uint16_t sw)
 {
     int ret = -1;
 
-    if (rdata != NULL) {
-        if (rdata->size - rdata->offset > IO_APDU_BUFFER_SIZE - 2 ||  //
-            !buffer_copy(rdata, G_io_apdu_buffer, sizeof(G_io_apdu_buffer))) {
-            return io_send_sw(SW_WRONG_RESPONSE_LENGTH);
+    if (rdatalist && count > 0) {
+        G_output_len = 0;
+        for (size_t i = 0; i < count; i++) {
+            const buffer_t *rdata = &rdatalist[i];
+
+            if (G_output_len + rdata->size - rdata->offset > IO_APDU_BUFFER_SIZE - 2
+                || !buffer_copy(rdata,
+                                G_io_apdu_buffer + G_output_len,
+                                sizeof(G_io_apdu_buffer) - G_output_len)) {
```

Comment:
This should be enough and less confusing, buffer_copy already check for reception buffer size.
```suggestion
            if (!buffer_copy(rdata,
                             G_io_apdu_buffer + G_output_len,
                             sizeof(G_io_apdu_buffer) - G_output_len - 2)) {
```
@yhql what do you think?

Reply:
I don't think that this change will work as intended because the final argument to buffer_copy() is size_t. size_t is an unsigned type which means that without an appropriate guard, if it underflows then it will become (on most implementations) a very large positive number and thus the check will not prevent a buffer overflow.

Reply:
I can't see how this could happen?
I we expand the buffer_copy check it gives us something like:

```
if (buffer->size - buffer->offset > sizeof(G_io_apdu_buffer) - G_output_len - 2) {
    return false;
}
```

if `buffer->size < buffer->offset` it will for sure underflow, but then it will be > `sizeof(G_io_apdu_buffer) - G_output_len - 2` which by construction can't underflow as `G_output_len` starts from `0` and is increased only if the condition above is respected.
Considering that `sizeof(G_io_apdu_buffer) - G_output_len - 2` won't underflow by construction, then in fact `buffer_copy` check is identic as the check I'm proposing to remove (`G_output_len + rdata->size - rdata->offset > IO_APDU_BUFFER_SIZE - 2`).

So I don't see how bad things could occurs.
Am I missing something here?

Reply:
When you call:
```
       buffer_copy(rdata,
                             G_io_apdu_buffer + G_output_len,
                             sizeof(G_io_apdu_buffer) - G_output_len - 2)
```
Let's say that G_output_len = sizeof(G_io_apdu_buffer) + 3. That argument is evaluated before being passed to the function. Using signed arithmetic, that would be -1. But, the variable is not signed and so it will end up being 0xffffffff which is coincidentally MAX_INT. When we get to the check in buffer_copy(), we see:
```
    if (buffer->size - buffer->offset > out_len)
```
our out_len in this example is 0xffffffff which can _never_ be true because that's MAX_INT.
Please keep in mind that I am assuming two compliment arithmetic and all that in this example. According to the C standard, the behaviour is undefined meaning that anything could happen.

Reply:
> Let's say that `G_output_len = sizeof(G_io_apdu_buffer) + 3`
This is not possible by construction, `G_output_len` max value is `sizeof(G_io_apdu_buffer) - 2`.
Else we have other issues, e.g.g already a bffer overflow on the previous buffer copy.

Reply:
@elric1 An up to date comment is always good!

Reply:
I force pushed a new version. I also dedup `G_output_len = 0;` because it doesn't need to be set in both branches of the conditional when it is perfectly acceptable to set it at the beginning.